prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -27,6 +29,152 @@ def check_for_location_conflicts(substations):
+    )
+
+    return line2sub
+
+
+def assign_substations_to_lines(substations, lines, line2sub):
+    """Locate closest substation of unconnected lines and update information
+    information of lines. Note that the ``line2sub`` data frame is modified in-place.
+
+    :param pandas.DataFrame substations: data frame of substations.
+    :param pandas.DataFrame lines: data frame of all lines.
+    :param pandas.DataFrame line2sub: indices: table mapping lines to substations
+        as returned by :func:`map_lines_to_substations_using_coords`.
+    :return: (*pandas.DataFrame*) -- original lines data frame with updated
+        *SUB_1* and *SUB_2* columns, new *SUB_1_ID*, *SUB_2_ID* along with
+        *OTHER_SUB_1_ID* and *OTHER_SUB_2_ID* when multiple location are
+        associated to the same endpoint.
+    """
+    print("finding closest neighbor to unconnected lines")
+
+    # find closest neighbor of unconnected lines
+    subcoord2subid = substations.groupby(["LONGITUDE", "LATITUDE"]).groups
+    subcoord = list(subcoord2subid)
+    end_sub = {"FROM": "FROM_SUB", "TO": "TO_SUB"}
+    missing_points = set().union(
+        *[set(line2sub[e].loc[line2sub[e_sub].isna()]) for e, e_sub in end_sub.items()]
+    )
+    endpoint2neighbor = {
+        point: subcoord2subid[subcoord[find_closest_neighbor(point, subcoord)]]
+        for point in tqdm(missing_points, total=len(missing_points))
+    }
+    filled_subs = [
+        line2sub[e].loc[line2sub[e_sub].isna()].map(endpoint2neighbor).map(list)
+        for e, e_sub in end_sub.items()
+    ]
+    line2sub.update(pd.concat(filled_subs, axis=1).rename(columns=end_sub))
+
+    # select all substations
+    new_substations = substations.copy()
+    new_substations = substations.loc[
+        set(pd.concat([line2sub[e] for e in ["FROM_SUB", "TO_SUB"]]).explode())
+    ]
+
+    # pick unique substation at each end. Order of preference:
+    # SUBSTATION --> TAP --> RISER --> DEAD END
+    all2one = {
+        tuple(i): None
+        for j in ["FROM_SUB", "TO_SUB"]
+        for i in line2sub[j]
+        if len(i) > 1
+    }
+    for a in all2one:
+        type2id = substations.loc[list(a)].reset_index().groupby("TYPE").first()["ID"]
+        for t in ["SUBSTATION", "TAP", "RISER", "DEAD END"]:
+            try:
+                all2one[a] = [type2id.loc[t]]
+                break
+            except KeyError:
+                continue
+
+    # Add column listing substations that were not picked
+    line2sub["OTHER_FROM_SUB"] = None
+    line2sub["OTHER_TO_SUB"] = None
+    for i in line2sub.index:
+        for endpoint in ["FROM", "TO"]:
+            if len(line2sub.loc[i, f"{endpoint}_SUB"]) > 1:
+                key = tuple(line2sub.loc[i, f"{endpoint}_SUB"])
+                line2sub.at[i, f"OTHER_{endpoint}_SUB"] = list(
+                    set(key) - set(all2one[key])
+                )
+                line2sub.at[i, f"{endpoint}_SUB"] = all2one[key]
```

Comment:
Not sure whether this is all we need here:
```python
type_rank = {"SUBSTATION": 1, "TAP": 2, "RISER": 3, "DEAD END": 4}
for c in ["FROM_SUB", "TO_SUB"]:
    line2sub.loc[:, c] = line2sub[c].apply(lambda x: sorted(x, key=lambda y: type_rank[substations.loc[y, "TYPE"]]))

```

Reply:
I like the idea. I would still need to save the other substations to populate the `OTHER_FROM_SUB` and `OTHER_TO_SUB` columns. I will see how I can do that with your implementation